tests/functionnal/CommonDBTM.php

```diff
@@ -329,4 +329,446 @@ function () use ($DB) {
+         'suppliers_id' => $ids[3]  // Child 3
+      ];
+      $this->boolean($rel->can(-1, CREATE, $input))->isTrue();
+      $idr[8] = (int)$rel->add($input);
+      $this->integer($idr[8])->isGreaterThan(0);
+      $this->boolean($rel->can($idr[8], READ))->isTrue();
+      $this->boolean($rel->canEdit($idr[8]))->isTrue();
+   }
+
+   /**
+    * Entity right check
+    */
+   public function testEntity() {
+      $this->login();
+
+      $ent0 = getItemByTypeName('Entity', '_test_root_entity', true);
+      $ent1 = getItemByTypeName('Entity', '_test_child_1', true);
+      $ent2 = getItemByTypeName('Entity', '_test_child_2', true);
+
+      $entity = new \Entity();
+      $ent3 = (int)$entity->add([
+         'name'         => '_test_child_2_subchild_1',
+         'entities_id'  => $ent2
+      ]);
+      $this->integer($ent3)->isGreaterThan(0);
+
+      $ent4 = (int)$entity->add([
+         'name'         => '_test_child_2_subchild_2',
+         'entities_id'  => $ent2
+      ]);
+      $this->integer($ent4)->isGreaterThan(0);
+
+      $this->boolean(\Session::changeActiveEntities('all'))->isTrue();
+
+      $this->boolean($entity->can(0, READ))->isTrue("Fail: can't read root entity");
+      $this->boolean($entity->can($ent0, READ))->isTrue("Fail: can't read entity 0");
+      $this->boolean($entity->can($ent1, READ))->isTrue("Fail: can't read entity 1");
+      $this->boolean($entity->can($ent2, READ))->isTrue("Fail: can't read entity 2");
+      $this->boolean($entity->can($ent3, READ))->isTrue("Fail: can't read entity 2.1");
+      $this->boolean($entity->can($ent4, READ))->isTrue("Fail: can't read entity 2.2");
+
+      $this->boolean($entity->canEdit(0))->isTrue("Fail: can't write root entity");
+      $this->boolean($entity->canEdit($ent0))->isTrue("Fail: can't write entity 0");
+      $this->boolean($entity->canEdit($ent1))->isTrue("Fail: can't write entity 1");
+      $this->boolean($entity->canEdit($ent2))->isTrue("Fail: can't write entity 2");
+      $this->boolean($entity->canEdit($ent3))->isTrue("Fail: can't write entity 2.1");
+      $this->boolean($entity->canEdit($ent4))->isTrue("Fail: can't write entity 2.2");
+
+      $input=['entities_id' => $ent1];
+      $this->boolean($entity->can(-1, CREATE, $input))->isTrue("Fail: can create entity in root");
+      $input=['entities_id' => $ent2];
+      $this->boolean($entity->can(-1, CREATE, $input))->isTrue("Fail: can't create entity in 2");
+      $input=['entities_id' => $ent3];
+      $this->boolean($entity->can(-1, CREATE, $input))->isTrue("Fail: can't create entity in 2.1");
+      $input=['entities_id' => 99999];
+      $this->boolean($entity->can(-1, CREATE, $input))->isFalse("Fail: can create entity in not existing entity");
+      $input=['entities_id' => -1];
+      $this->boolean($entity->can(-1, CREATE, $input))->isFalse("Fail: can create entity in not existing entity");
+
+      $this->boolean(\Session::changeActiveEntities('all'))->isTrue();
```

Comment:
Looks like there is a lot of duplication in tests. From this line to line 750, only tests on `99999` are not duplicated (see lines 707 to 721).

Reply:
Lines from 752 to 761 are duplications too.